lib/perf_hooks.js

```diff
@@ -0,0 +1,370 @@
+    case 'gc': return NODE_PERFORMANCE_ENTRY_TYPE_GC;
+    case 'function': return NODE_PERFORMANCE_ENTRY_TYPE_FUNCTION;
+  }
+}
+
+class PerformanceEntryList {
+  constructor(entries) {
+    this[kEntries] = new Set(entries);
+
+    Object.defineProperties(this, {
+      [kEntries]: { writable: true, enumerable: false }
+    });
+  }
+
+  getEntries() {
+    return Array.from(this[kEntries]);
+  }
+
+  getEntriesByType(type) {
+    type = `${type}`;
+    return this.getEntries().filter((entry) => {
+      return entry.entryType === type;
+    });
+  }
+
+  getEntriesByName(name, type) {
+    name = `${name}`;
+    type = type !== undefined ? `${type}` : type;
+    return this.getEntries().filter((entry) => {
+      return type === undefined ?
+        entry.name === name :
+        entry.name === name && entry.entryType === type;
+    });
+  }
+}
+
+class PerformanceObserver {
+  constructor(callback) {
+    if (typeof callback !== 'function') {
+      const errors = lazyErrors();
+      throw new errors.TypeError('ERR_INVALID_CALLBACK');
+    }
+    this[kCallback] = callback;
+    this[kBuffer] = new PerformanceEntryList([]);
+
+    Object.defineProperties(this, {
+      [kTypes]: { enumerable: false, writable: true },
+      [kCallback]: { enumerable: false, writable: true },
+      [kBuffer]: { enumerable: false, writable: true },
+      [kBuffering]: { enumerable: false, writable: true },
+      [kQueued]: { enumerable: false, writable: true }
+    });
+  }
+
+  disconnect() {
+    const types = this[kTypes];
+    if (types === undefined || types.length === 0)
+      return;
+    for (var n = 0; n < types.length; n++) {
+      const set = getObserversSet(types[n]);
+      set.delete(this);
+      observerCounts[types[n]] = set.size;
+    }
+  }
+
+  observe(options) {
+    const errors = lazyErrors();
+    if (typeof options !== 'object' || options == null) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'options', 'Object');
+    }
+    if (!Array.isArray(options.entryTypes)) {
+      throw new errors.TypeError('ERR_INVALID_OPT_VALUE',
+                                 'entryTypes', options);
+    }
+    const entryTypes = options.entryTypes.filter(filterTypes).map(mapTypes);
+    this[kBuffer][kEntries].clear();
+    this[kBuffering] = Boolean(options.buffered);
+    if (entryTypes.length === 0) {
+      throw new errors.Error('ERR_VALID_PERFORMANCE_ENTRY_TYPE');
+    }
+    this[kTypes] = entryTypes;
+    for (var n = 0; n < entryTypes.length; n++) {
+      const set = getObserversSet(entryTypes[n]);
+      set.add(this);
+      observerCounts[entryTypes[n]] = set.size;
+    }
+  }
+}
+
+class Performance {
+  get nodeTiming() {
+    return nodeTiming;
+  }
+
+  get nodeFrame() {
+    return nodeFrame;
+  }
+
+  get timeOrigin() {
+    return timeOrigin;
+  }
+
+  now() {
+    const hr = process.hrtime(timeOrigin);
+    return hr[0] * 1000 + hr[1] / 1e6;
+  }
+
+  getEntries() {
+    const ret = getEntries();
```

Comment:
The naming conflict here seems a bit confusing...can we make this `getEntriesFromTimeline` or something different?